### PR TITLE
allow variable grain size in diffusion creep

### PR DESCRIPTION
**Pull request by bobmyhill**
This PR adds grain size as a variable argument to the functions in DiffusionCreep. Each function is overloaded so that old functionality is preserved. 

The old functions call the new functions, using the fixed grain size as input.

The changes in this PR will be useful for composite rheologies and material models where grain size is a variable.

* [x] I have followed the [instructions for indenting my code](../blob/main/CONTRIBUTING.md#making-aspect-better).
* [x] I have tested my new feature locally to ensure it is correct.
* [x] I have added a changelog entry in the [doc/modules/changes](../blob/main/doc/modules/changes) directory that will inform other users of my change.

### CODE REVIEW

**Comment by bobmyhill** [comment]
It's not obvious to me whether the compiler will inline the wrapped functions. Even if the answer is no, perhaps the overhead is small enough that we don't care?

File: include/aspect/material_model/rheology/diffusion_creep.h [hunk 1]
```diff
@@ -159,7 +198,7 @@ namespace aspect
           /**
            * Diffusion creep grain size d.
```
Comment (gassmoeller) [inline include/aspect/material_model/rheology/diffusion_creep.h:199]: Please extend this to explain that  this is read from the input file, and only used by the functions that do not take the grain size as additional argument.

File: source/material_model/rheology/diffusion_creep.cc [hunk 2]
```diff
@@ -270,7 +311,7 @@ namespace aspect
         activation_volumes_diffusion = Utilities::MapParsing::parse_map_to_double_array(prm.get("Activation volumes for diffusion creep"),
                                                                                         options);
 
-        grain_size = prm.get_double("Grain size");
+        fixed_grain_size = prm.get_double("Grain size");
```
Comment (gassmoeller) [inline source/material_model/rheology/diffusion_creep.cc:314]: Could you extend the documentation of this parameter in `declare_parameters` above to say that this grain size is only being used if the active material model does not provide its own (possibly changing grain size) when calling this rheology?

**Review by gassmoeller — COMMENTED** [review]
Yes, this is probably the approach that maintains the highest degree of compatibility and flexibility. I would not worry about the cost of function calls, unless it shows up in a profile later, it is almost always negligible (I bet both the `std::pow` and `std::exp` inside the function is one to two orders of magnitude more expensive).

Could you fix the comments as requested?

**Comment by bobmyhill** [comment]
Done! Thanks @gassmoeller :)

File: source/material_model/rheology/diffusion_creep.cc [hunk 3]
```diff
@@ -257,6 +257,11 @@ namespace aspect
                            "If only one value is given, then all use the same value. "
                            "Units: \\si{\\meter\\cubed\\per\\mole}.");
         prm.declare_entry ("Grain size", "1e-3", Patterns::Double (0.),
+                           "The fixed grain size of the material. "
+                           "This grain size is only used if the parent "
+                           "material model does not provide its own "
+                           "(possibly variable grain size) when "
```
Comment (gassmoeller) [inline source/material_model/rheology/diffusion_creep.cc:263]:
```suggestion
                           "(possibly variable) grain size when "
```

**Review by gassmoeller — APPROVED** [review]
Almost ;-). Good to go when the typo is fixed.

**Pull request merged**